### PR TITLE
make substr, lsh and ash parameters be int32

### PR DESCRIPTION
**Pull request by arvidn**
These are the last places where we currently don't charge for redundant leading zeros. ban wider than 4 bytes, even if they are leading zeros. This avoid having to charge cost for such extra bytes.

### CODE REVIEW

**Comment by arvidn** [comment]
I actually meant to update the unknown op cost calculation as well, to use this same restriction. (the same way I did in the rust implementation [here](https://github.com/Chia-Network/clvm_rs/pull/52/files#diff-1106ab16843af039b2910d14818b53d52ab8542371d26e13f95b080b44617934L135))

File: clvm/operators.py [hunk 1]
```diff
@@ -111,6 +111,10 @@ def default_unknown_op(op: bytes, args: CLVMObject) -> Tuple[int, CLVMObject]:
 
     cost_function = (op[-1] & 0b11000000) >> 6
     # the multiplier cannot be 0. it starts at 1
+
+    if len(op) > 5:
+        raise EvalError("invalid operator", args.to(op))
```
Comment (arvidn) [inline clvm/operators.py:116]: the new requirements is that the cost multiplier may not be more than 4 bytes wide. It's still interpreted as an unsigned integer. This restriction was kind of implied before, by ensuring the total cost of the operation did not exceed `uint32_max`.

File: tests/unknown-op/unknown-00.txt [hunk 2]
```diff
@@ -1,2 +1,2 @@
-brun --strict -c '(a (q 0x0000) (q ()))'
+brun --strict -c '(a (q . 0x000) (q . ()))'
```
Comment (arvidn) [inline tests/unknown-op/unknown-00.txt:1]: I'm not sure why these worked without having the quote updated. But I re-generated these to cover more interesting cases, with the maximum cost multiplier as well as invalid cost multipliers (5 bytes wide)

**Comment by arvidn** [comment]
this was the program to re-generate the unknown op test cases:
```
cnt = 0
for args in ['(q . ())', '(q . (1 2 3))', '(q . (0x101010101010101010101010101010101010101010101010101010101010101010101010101010101010101101010101010101010101010101010101010101010101010101010101010101010101010101010101010101 0x101010101))', '(q . (1 2 (1 10 20)))']:
        for cost_fun in [0, 1, 2, 3]:
                for multiplier in [0, 1, 0xfffeffff, 0x7ffffffff]:
                        for ignored in [0, 63]:
                                for strict in [True, False]:
                                        open("unknown-%02d.txt" % cnt, "w+").write("brun " + ("--strict " if strict else "") + "-c '(a (q . 0x%x%02x) %s)'" % (multiplier, (cost_fun << 6) | ignored, args))
                                        cnt += 1

```

File: tests/unknown-op/unknown-00.txt (hunk 2 above)
Comment (richardkiss) [reply tests/unknown-op/unknown-00.txt:1]: If you're changing these, can you change them to not use `a` (since its semantics are changing)? You can just go `(0x0000 (q . ()))` for example.

File: clvm/more_ops.py [hunk 3]
```diff
@@ -68,6 +68,15 @@ def args_as_ints(op_name, args):
         yield (arg.as_int(), len(arg.as_atom()))
 
 
+def args_as_int32(op_name, args):
```
Comment (richardkiss) [inline clvm/more_ops.py:71]: What do you think about moving this guy to `casts.py`?

Comment (richardkiss) [reply clvm/more_ops.py:71]: Oh I guess that doesn't make sense, nm.

File: clvm/operators.py (hunk 1 above)
Comment (richardkiss) [inline clvm/operators.py:116]: We could just reset the cost in this case to something small, even a fixed value, rather than throw out the operator. I dunno, probably doesn't much matter. What do you think?

**Review by richardkiss — APPROVED** [review]
Looks great! The changes I suggested are optional, use your judgement as guide.

File: clvm/more_ops.py [hunk 4]
```diff
@@ -261,7 +272,9 @@ def op_ash(args):
 
 
 def op_lsh(args):
-    (i0, l0), (i1, _) = args_as_int_list("lsh", args, 2)
+    (i0, l0), (i1, l1) = args_as_int_list("lsh", args, 2)
```
[the remark this answers is not in the document]
Comment (arvidn) [reply clvm/more_ops.py:275]: the thing here is that the first parameter needs to be a full `BigInt`. It's just the second one that's an `int32`

File: clvm/operators.py (hunk 1 above)
Comment (arvidn) [reply clvm/operators.py:116]: I would think we have enough space for additional operators. If we feel we should have more space, I would think the lower bits (that are ignored for cost purposes) should be expanded.

**Comment by arvidn** [comment]
I updated the unknown-operator tests to not use apply

File: clvm/more_ops.py [hunk 5]
```diff
@@ -224,7 +233,7 @@ def op_substr(args):
     if a0.pair:
         raise EvalError("substr on list", a0)
 
-    (i1, _), (i2, _) = args_as_int_list("substr", args.rest(), 2)
+    i1, i2 = list(args_as_int32("substr", args.rest()))
```
Comment (aqk) [inline clvm/more_ops.py:236]: Ah, good idea.

File: clvm/more_ops.py (hunk 3 above)
Comment (aqk) [reply clvm/more_ops.py:71]: `args_as_int32` could take an arg_count param

File: clvm/more_ops.py [hunk 6]
```diff
@@ -248,7 +257,9 @@ def op_concat(args):
 
 
 def op_ash(args):
-    (i0, l0), (i1, _) = args_as_int_list("ash", args, 2)
+    (i0, l0), (i1, l1) = args_as_int_list("ash", args, 2)
+    if l1 > 4:
+        raise EvalError("ash requires int32 args (with no leading zeros)", args.rest().first())
```
Comment (aqk) [inline clvm/more_ops.py:262]: I understand you can't use `args_as_int32` on the full arg list, but you might be able to centralize this check by:

>(i0, l0), (i1, l1) = args_as_int_list("ash", args, 2)
> i1 = args_as_int32("ash", args.rest(), 1)

Also redundant, unfortunately - No big deal
:)

File: clvm/operators.py [hunk 7]
```diff
@@ -77,15 +77,15 @@ def args_len(op_name, args):
...
-# n | .... | 1          | 0          |
-# --+- - --+------------+------------+
-# n | .... |            |XX | XXXXXX |
-# --+- - --+------------+---+--------+
-# ^                      ^   ^
-# |                      |   + 6 bits ignored when computing cost
-# cost_multiplier        |
-#                        + 2 bits
-#                          cost_function
+# | 4 | 3 | 2 | 1 | 0          |
```
Comment (aqk) [inline clvm/operators.py:80]: Oh, that's more readable, thanks!

File: clvm/more_ops.py (hunk 3 above)
Comment (arvidn) [reply clvm/more_ops.py:71]: note that this is actually a generator, so it would be a complicated place to check number of arguments. However, this is only used after the number of arguments are already checked

File: clvm/operators.py (hunk 1 above)
Comment (aqk) [reply clvm/operators.py:116]: Can you add the comment that it's still interpreted as an unsigned integer?
Maybe that is discussed during cost computation.

File: tests/brun/ash-c.txt [hunk 8]
```diff
@@ -1,3 +1,2 @@
 brun -c '(ash (q . 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007) (q . 0x0000000000000000000000000000000000000000000000000000000000000001))'
-cost = 76
-14
+FAIL: ash requires int32 args (with no leading zeros) 0x0000000000000000000000000000000000000000000000000000000000000001
```
Comment (aqk) [inline tests/brun/ash-c.txt:2]: These new size restrictions for args 2&3 of substr and arg 2 of ash & lsh are meant to apply in all modes (safe & develop), right?

That seems correct to me.

File: tests/brun/lsh-c.txt [hunk 9]
```diff
@@ -1,3 +1,2 @@
 brun -c '(lsh (q . 0x00000000000000000000000000000000000000000000000000000000000000000007) (q . 0x000000000000000000000000000000000000000000000000000000000000000000001))'
-cost = 76
-14
+FAIL: lsh requires int32 args (with no leading zeros) 0x0000000000000000000000000000000000000000000000000000000000000000000001
```
Comment (aqk) [inline tests/brun/lsh-c.txt:2]: Hmm, now that these fail because of arg size, we need new tests on the boundary.

File: clvm/operators.py (hunk 1 above)
Comment (aqk) [reply clvm/operators.py:116]: Can you add the comment that it's still interpreted as an unsigned integer?
Maybe that is discussed during cost computation.

File: tests/brun/ash-c.txt (hunk 8 above)
Comment (arvidn) [reply tests/brun/ash-c.txt:2]: that's right

File: tests/brun/lsh-c.txt (hunk 9 above)
Comment (arvidn) [reply tests/brun/lsh-c.txt:2]: good point, I'll add those (in a new PR)

File: clvm/more_ops.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (richardkiss) [reply clvm/more_ops.py:275]: Oh yeah, ugh!
